bsc.evaluator/undefined/undefined.exp

```diff
@@ -6,10 +6,8 @@ compile_verilog_fail_error UndefinedStringErr.bsv S0015
 compile_object_fail_error UndefinedStringLength.bsv G0080
 
 if {$ctest == 1} {
-  compile_object_fail_error AbstractDeriveType.bsv S0015
-  compare_file AbstractDeriveType.bsv.bsc-ccomp-out
-  compile_object_fail_error AbstractDerivePosition.bsv S0015
-  compare_file AbstractDerivePosition.bsv.bsc-ccomp-out
+  compile_verilog_pass AbstractDeriveType.bsv mkAbstractDeriveType
```

Comment:
See the comment here: https://github.com/krame505/bsc/blob/generics/src/Libraries/Base1/Prelude.bs#L3909

Basically, we can't rely on the generic implementation of `PrimMakeUndefined` for `Position__` and `Type` since these are used in building error messages, which lead to infinite recursion.  For `Position__` we decided to make this just the position of the undefined value being demanded, and for `Type` we made this be `Type` (this was @nanavati's idea; since these are "don't care" values, giving "real" values in these cases presumably shouldn't pose an issue.)  If you have a better idea for what these defaults should be, I am definitely open to considering them.

Reply:
I don't understand two things:

For one, why is this a new problem that has been introduced?  BSC was generating instances for primitive types before, and it wasn't leading to infinite recursion, so why is that happening now?

For two, why can't the same definition (that BSC was generating before) be used now (just given explicitly and not autogenerated)?  BSC was generating instances for primitive types using this:
```
doPrimTypeUndefined :: Id -> [CType] -> CDefn
doPrimTypeUndefined i vs = Cinstance (CQType [] (TAp (cTCon idUndefined) ty)) [undef]
  where ty  = cTApplys (cTCon i) vs
        aty = tPosition `fn` tInteger `fn` ty
        undef = CLValueSign (CDef idMakeUndefinedNQ (CQType [] aty) [CClause [CPVar id_x, CPVar id_y] [] body]) []
        body = cVApply idPrimError [CVar id_x, str_expr]
        str_expr = cVApply idPreludePlus [error_str, type_str]
        error_str = stringLiteralAt noPosition "Attempt to use undetermined "
        type_str = cVApply idPrintType [typeLiteral ty]
```
That is, why can't you write this instance in the Prelude:
```
instance PrimMakeUndefined Type where
  primMakeUndefined pos _ = primError pos "Attempt to use undetermined Type"
```

In this explicit instance, there's no need to call `printType` because we know the type, so we can just include it in the string literal.  Is this the use of the type that you felt was recursion?  Even if it was kept it, it's not recursion on the typeclass.  We do use the position that was passed in, but again, that's not recursion of the typeclass and shouldn't cause problems.  I imagine that we could define the same explicit instance for `Position__` (although I'd want to confirm that this what BSC was generating before):
```
instance PrimMakeUndefined Position__ where
  primMakeUndefined pos _ = primError pos "Attempt to use undetermined Position__"
```

Reply:
From what I remember, the ConcPrim instance of PrimMakeUndefined ends up using an undefined Type and Position__ as part of its plumbing. I don't think we fully figured out why (Constructing a placeholder value "returned" by PrimError? Required to build or traverse the generic metadata?), but that's why the falling back on the generic ConcPrim instance is a non-starter - if raising an error about using an undefined value of a primitive type needs an undefined value of the primitive type Type then ...

We also tried the error-raising versions of those instances. I think the problem there is that these undefined values were demanded at an inconvenient time. So if you used an undefined value of a primitive type, it would want an undefined value of Type Type and you'd get an error about that (or Position__ after we special-cased Type) instead of an error about the thing you actually used. Or, depending on implementation details, you could get on the undefined Type or Position merry-go-round from a different starting point.

@krame505 can correct me if I'm wrong about this, but I think we also did the experiment of just not telling you the type of the primitive value you just misused. In that case, I think error-raising instances for Type and Position__ were fine (not sure if we tried removing them entirely), but that's a bad tradeoff. I think that experiment was also what convinced me that the undefined type and position were unavoidable (if you want good error messages for the other primitives).

Reply:
The details are a bit fuzzy for me as well, but I think that is generally the situation.  For what it's worth though, I just tried changing those instances to what Julie suggested and the test suite still passes, aside from the AbstractDerive* tests (whose expected output we had changed to match this behavior.)  So whatever the issue was with calling `primError`, this seems to have been avoided by later changes?

Reply:
If I remove the explicit instances for `Type` and `Position__` (not just replace them with my definitions) that I can successfully compile the libraries and run the `bsc.generics`, `bsc.typechecker`, and `bsc.evaluator/undefined/` tests.  I assume that I would have seen the problem, or do you recall that the issue showed up somewhere else?

I see this in the Prelude:
```
-- Default generic instance covers data and structs
 instance (Generic a r, PrimMakeUndefined' a r) => PrimMakeUndefined a where
   primMakeUndefined pos i = primMakeUndefined' ((error "proxy") :: r) pos i

 instance (PrimMakeUndefined a) => PrimMakeUndefined' a (ConcPrim a) where
   primMakeUndefined' (ConcPrim x) pos _ = primError pos $
     "Attempt to use undetermined " +++ (printType (typeOf x))
```
It does seem odd to me that the `primMakeUndefined'` function is matching against its first argument, which is created with `error`.  I might want to clean that up, but it seems to be working for now.  The instance also has a `PrimMakeUndefined` context, which doesn't hurt but is presumably not needed, so I wonder if you at one time you had a different definition (that was using a don't-care).  I'll look through the commit history.

Reply:
I do see that the `ConcPrim` instance of `PrimMakeUndefined'` had previously been an instance of `PrimMakeUndefined''` -- it was moved in commit `1c4e1154`.  If I revert that, then the existence of the explicit instances for `Type` and `Position__` matter.  Without them, the test `AbstractDeriveType.bsv` gives:
```
Internal Bluespec Compiler Error:
evalHeap - prim should be evaluated:(PrimPrintType, PrimPrintType _ ::  Prelude.Type)
```
Is that what you recall seeing?

Reply:
The prior commit (`fdaa77cb`), that added the explicit instances for `Type` and `Position__`, also changed that `PrimMakeUndefined'` instance for `ConcPrim` to not use a don't care:
```
-    "Attempt to use undetermined " +++ (printType (typeOf (_ :: a)))
+    "Attempt to use undetermined " +++ (printType (typeOf ((error "proxy") :: a)))
```
That change may have affected the `Position__` test.

Anyway... It seems to me that it's now safe to remove the explicit instances for `Type` and `Position__`.

And also I think it'd be cleaner to remove the matching on the first argument (used to get `x` for printing the type) and go back to creating a value of type `a` and using that.  I find that the following definition passes the testsuite (although replacing `_` with `error "proxy"` is also fine):
```
instance (PrimMakeUndefined a) => PrimMakeUndefined' a (ConcPrim a) where
  primMakeUndefined' _ pos _ = primError pos $
    "Attempt to use undetermined " +++ (printType (typeOf (_ :: a)))
```